Add shelley tx support

paweljakubas opened the pull request:
#16 

Done:
- [x] I have updated dependencies
- [x] I have add Shelley scaffolding
- [x] I have add mkInput
- [x] I have make new place for `from*` for better reuse later
- [x] I have add mkOutput
- [x] I have add signWith for shelley witnesses
- [x] I have add signWith for byron witnesses
- [x] I have add lock
- [x] I have add serialise
- [x] I have added more payment errors 
- [x] I have handled fee estimation/calculation
- [x] I have add tests
- [x] I have added golden tests

paweljakubas committed:
bump dependencies and add Shelley network constructors

paweljakubas committed:
use package.yaml rather than deirectly cabal

paweljakubas committed:
base TxIn and TxOut on cardano-api types

paweljakubas committed:
add mkInput

paweljakubas committed:
add several MkPayment function impls

paweljakubas committed:
add helpers for decoding

paweljakubas committed:
add mkOutput

paweljakubas committed:
add lock impl and correct init

paweljakubas committed:
add witnesses list to Tx Shelley

paweljakubas committed:
mkShelleySigningKey

paweljakubas committed:
add signWith impl for shelley witness

paweljakubas committed:
add byron witnesses support to shelley tx

paweljakubas committed:
add serialize impl

paweljakubas committed:
add too low output error

paweljakubas committed:
propose a way to estimateFee using cardano-api primitives

paweljakubas committed:
update serialization - use the same method like for shelley unit testing

paweljakubas committed:
improve on estimateFee

paweljakubas committed:
add first tests for shelley plus fine-tune mkOutput to omit ScriptHashObj credentials

paweljakubas committed:
rethink fee estimation

paweljakubas committed:
correct invalid output generator

paweljakubas committed:
add blob in integration testing plus correct shelley witnesses construction

KtorZ commented on src/Data/UTxO/Transaction/Cardano/Helpers.hs:
```diff
@@ -0,0 +1,79 @@
+{-# OPTIONS_HADDOCK prune #-}
+{-# OPTIONS_GHC -Wno-unrecognised-pragmas #-} -- needed for {#- HLINT ... #-}
+
+module Data.UTxO.Transaction.Cardano.Helpers
```
Having these as a separate module makes sense, but these have nothing to do with UTxO nor cardano ;)

paweljakubas replied:
this is probably temporary in the end as we will remove Byron stuff, and this could be merged with Shelley.

KtorZ commented on src/Data/UTxO/Transaction.hs:
```diff
@@ -125,6 +125,8 @@ data ErrMkPayment
     | MissingSignature
         -- ^ Payments must have a signature for each input.
+    | TooLowOutput
+        -- ^ Each payment must not be smaller than minUTxOvalue
```
Perhaps `TooSmallOutput` ?

KtorZ replied:
Although, how would we know that from cardano-transactions :thinking:  ? This depends on protocol parameters which this library is not aware of.

paweljakubas replied:
if we do not set this here, then if someone constructs output which is  < minUTxOvalue and use external tx endpoint then node will reject it and 500 will be emitted (at this moment). So I decided to introduce it here. But I agree it is not so straightfoward

KtorZ replied:
> then node will reject it and 500 will be emitted (at this moment)

Indeed and that is totally fine. We could make the error reporting on cardano-wallet better, but cardano-transactions isn't the place to fix this :+1:

paweljakubas replied:
agree, handling/exposing node errors in for example external tx endpoint is sound approach. reverted this change in https://github.com/input-output-hk/cardano-transactions/pull/17/commits/51ce57e3a1225b95303143a429a4a4d98b9a501b

KtorZ commented on src/Data/UTxO/Transaction/Cardano/Shelley.hs:
```diff
@@ -0,0 +1,386 @@
+        -- ^ A ttl expressed in slot number counted from the beginning of blockchain
+    -> Word64
+        -- ^ fee of tx as taken when constructing change outputs
+        -- see also estimateFee
```
Hmmm. This last parameter looks wrong. Fee estimation should not happen at this level. This library is simply for constructing transactions. It assumes nothing about fee estimation.

paweljakubas replied:
the reason is this https://github.com/input-output-hk/cardano-node/blob/master/cardano-api/src/Cardano/Api/Typed.hs#L1040

paweljakubas replied:
TxFee determines TxBody Shelley and this is signed .... So I had several attempts to address fee (if you look at commit history)

KtorZ replied:
I see (and forgot about the explicit fee in the Shelley transactions for a moment). Okay, I'll look a bit more into this function which I sort of skip during the first review :+1:

KtorZ commented on src/Data/UTxO/Transaction/Cardano/Shelley.hs:
```diff
@@ -0,0 +1,386 @@
+    lock (_net, _ttl, _fee, [], _outs) = Left MissingInput
+    lock (_net, _ttl, _fee, _inps, []) = Left MissingOutput
+    lock (net, ttl, fee, inps, outs)
+        | any (\(TxOut _ coin) -> coin < minUTxOvalue) outs = Left TooLowOutput
```
This value depends on protocol parameters, we can't possibly know that in this library. The caller has to be aware. This library should generate transaction that are structurally well-formed, but they may still be interpreted as invalid if the network deems them so. Very much we can't verify that the provided UTxO exists or has the right amount, we can't also verify the minimum UTxO amount.

paweljakubas replied:
fixed in https://github.com/input-output-hk/cardano-transactions/pull/17/commits/51ce57e3a1225b95303143a429a4a4d98b9a501b

KtorZ commented on src/Data/UTxO/Transaction/Cardano/Shelley.hs:
```diff
@@ -0,0 +1,386 @@
+            TxExtraContent
+                { txMetadata = Nothing
+                , txWithdrawals = []
+                , txCertificates = []
```
Okay for now, but the library should support adding certificates and withdrawals, as well as metadata.

paweljakubas replied:
yes, this is supposed to be the first version. My intention was to have first version working with extenral tx endpoint of cardano-wallet. In the next versions we should extend by certs, withdrawals, metadata

KtorZ replied:
Okay, I'll suggest leaving a `TODO` mark in the source code then which is linked to a corresponding github issue to make sure this doesn't get forgotten.

paweljakubas committed:
fix 2 bugs that allows proper signing with byron witnesses

paweljakubas committed:
remove dealing with minUTxO value

paweljakubas committed:
better xprvFromBytes reuse and first golden test

paweljakubas committed:
add gloden tests for shelley witnesses

paweljakubas committed:
add gloden tests for byron witnesses

KtorZ commented on src/Data/UTxO/Transaction/Cardano/Shelley.hs:
```diff
@@ -0,0 +1,373 @@
+
+      -- Values to calculate fee : b + a*txsize
+      minFeeA = 44
+      minFeeB = 155381
```
These values only hold for current Mainnet (and may even be different tomorrow). I still don't fully understand the need for a _fee estimation_ function since we are constructing the actual transaction and thus, all inputs and outputs are known. There's no need to estimate.

paweljakubas replied:
I removed feeEstimation - I agree it would be better to service that in external tx one day - https://github.com/input-output-hk/cardano-transactions/pull/17/commits/23deb590347fdd28c2f99a2f008ab42c098cead8

paweljakubas committed:
update to 1.19.1 and remove estimateFee

KtorZ committed:
rename 'Data.UTxO.Transaction.Cardano.Helpers' as 'Codec.Binary.Encoding'

KtorZ committed:
move payload deserialization in its own module and make the API a bit more type safe

  This caused some of the tests to fail (with reason!). The API was
  accepting any bytestring as a valid address and simply assuming that
  there was no derivation path in the given address. The new API is a
  bit more subtle and make sure that the given 'ByteString' given to
  sign a byron get does represent a valid Byron address. So, I adjusted
  the tests to use well-formed byron addresses instead of arbitrary
  bytestrings.

KtorZ committed:
disable systemd on cardano-config

  - can't be statically linked
  - irrelevant for cardano-transactions
